src/coreclr/System.Private.CoreLib/src/System/ValueType.cs

```diff
@@ -79,8 +95,44 @@ public override unsafe bool Equals([NotNullWhen(true)] object? obj)
         **Arguments: None.
         **Exceptions: None.
         ==============================================================================*/
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        public extern override int GetHashCode();
+        public override unsafe int GetHashCode()
+        {
+            // The default implementation of GetHashCode() for all value types.
+            // Note that this implementation reveals the value of the fields.
+            // So if the value type contains any sensitive information it should
+            // implement its own GetHashCode().
+
+            MethodTable* pMT = RuntimeHelpers.GetMethodTable(this);
+
+            // We don't want to expose the method table pointer in the hash code
+            // Let's use the typeID instead.
+            uint typeID = RuntimeHelpers.GetTypeID(pMT);
+
+            // To get less colliding and more evenly distributed hash codes,
+            // we munge the class index with two big prime numbers
+            int hashCode = (int)(typeID * 711650207 + 2506965631U);
+
+            if (CanCompareBitsOrUseFastGetHashCode(pMT))
+            {
+                // this is a struct with no refs and no "strange" offsets
+                HashCode hash = default;
+                uint size = pMT->GetNumInstanceFieldBytes();
+                hash.AddBytes(MemoryMarshal.CreateReadOnlySpan(ref this.GetRawData(), (int)size));
+                hashCode ^= hash.ToHashCode();
+            }
+            else
+            {
+                object obj = this;
+                hashCode ^= RegularGetValueTypeHashCode(pMT, ObjectHandleOnStack.Create(ref obj));
+            }
+
+            GC.KeepAlive(this);
+            return hashCode;
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ValueType_RegularGetValueTypeHashCode")]
+        [SuppressGCTransition]
```

Comment:
The unmanaged slow path manipulates the object reference in COOP mode

Reply:
It can also throw and call back into managed code. It is not ok to use SuppressGCTransition with that.

SuppressGCTransition can be only used for leaf methods.

Reply:
If you would like to make it more managed, you can change the helper to return the strategy to use to compute the hashcode:

- Call object.GetHashCode for object reference at given offset
- Call double.GetHashCode for field at given offset 
- Call float.GetHashCode for field at given offset
- Hash N bytes from given offset (this can also cover the case where `CanCompareBitsOrUseFastGetHashCode` was not computed)

Reply:
Got some chanllenge when handling the recursive case.
How to pass a reference pointing into an object to QCall? There's a leaf method `GetFieldTypeHandleThrowing` marked as throwing, so it looks unsuitable for FCall since we are moving away from HELPER_METHOD_FRAME.
Or, will it really throw when loading a field handle of already boxed type? Is there any alternate?

Reply:
Well, P/Invoke generator can pin the `ref byte`. Is it reliable/performant for object fields? Can GCPROTECT be omitted then?

Reply:
I think it would work better if you return field offset from the QCall and compute the byref on the managed side. Nothing to GC protect, etc.

Reply:
For the recursive value type case, it requires a reference pointing to the field, which is not a top level object. Passing object handle+offset does work, but looks a bit confusing if more levels are involved.

Reply:
Compound offset like this is not uncommon. It is frequently used e.g. by JITed code.

My preference is to have less manually managed code. Less GC mode switches in VM and less manually managed code that runs in a cooperative mode is goodness.